Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,174 @@
+the configuration.
+
+Observe we could move some operations to the `HashCache<?>` instead, for example, `remove(K)`, `containsKey(K)`,
+and `clear()`. Another possibility is to provide these at the Hash and HashCache level, so the user can apply
+operations without transferring the complete `Hash<?, ?>`. It would be especially useful when the Hash state
+changes, so the user does not need to pass the structure back and forth.
+
+[IMPORTANT]
+.Question?
+====
+How do we deal with concurrent updates to the same Hash property?
+====
+
+To interact with the `Hash<K, V>`:
+
+[source, java]
+.HashCache.java
+----
+public interface HashCache<K> {
+
+   <HK, HV> HV put(K key, HK hashKey, HV hashValue);
+
+   <HK, HV> Hash<HK, HV> put(K key, Hash<HK, HV> hash);
+
+   <HK, HV> HV get(K key, HK hashKey); // <1>
+
+   Hash<HK, HV> get(K key);
+
+   boolean remove(K key);
+
+   <HK> boolean remove(K key, HK hashKey);
+
+   <HK, HV> void remove(Predicate<Hash<HK, HV>> p);
+}
+----
+
+<1> Do we provide a varags version of this method? To retrieve only a subset of keys.
+
+And as of Multimap, we provide APIs for:
+
+=== Hot Rod API
+
+The new API will be developed. Provide implementations for an `AsyncHashCache` and `MutinyHashCache`.
+
+=== REST API
+
+Hash REST API should be the almost the same as the Cache one.
+The difference is in the endpoint.
+
+Examples:
+[source]
+----
+GET /rest/v2/hashes/
+
+GET /rest/v2/hashes/{hashName}?action=???
+
+DELETE /rest/v2/hashes/{hashName}/{hashKey}
+
+DELETE /rest/v2/hashes/{hashName}/{hashKey}/{hashProperty}
+----
+
+* CRUD hashes (providing a configuration to create a hash, templates not supported)
+* Display Hash content entries
+* Get values by key
+* In Indexing is enabled, hashes support query
+* CRUD for operations available in the `HashCache<?>` API
+* Enable/Disable rebalancing
+* Availability get/set
+* Replace not supported
+* Listeners support to enable interactions with near-caching
+
+
+== Using the Hash
+
+A new maven module is created.
+
+=== Configuration Schema
+
+[source,xml]
+----
+<xs:complexType name="hashes">
+    <xs:sequence>
+        <xs:element name="hash" type="hash" minOccurs="0" maxOccurs="unbounded">
+            <xs:annotation>
+                <xs:documentation>
+                    Defines a hash data structure.
+                </xs:documentation>
+            </xs:annotation>
+        </xs:element>
+    </xs:sequence>
+</xs:complexType>
+
+<xs:complexType name="hash">
+    <xs:attribute name="name" type="xs:ID" use="required">
+        <xs:annotation>
+            <xs:documentation>
+                A unique name for the hash.
+            </xs:documentation>
+        </xs:annotation>
+    </xs:attribute>
+</xs:complexType>
+----
+
+To create and use a `HashCache<?>`:
+
+```java
+      EmbeddedCacheManager cm = ...
+      HashCacheManager hashCacheManager = EmbeddedHashCacheManagerFactory.from(cm);
```

Comment:
this is the legacy API, we need this in the new API only

Reply:
Let's not implement anything on the legacy API please.